lib/util/exec/buildpack.spec.ts

```diff
@@ -127,14 +118,38 @@ describe('util/exec/buildpack', () => {
     });
 
     it('falls back to latest version if invalid constraint', async () => {
-      datasource.getPkgReleases.mockReset();
       datasource.getPkgReleases.mockResolvedValueOnce({
         releases: [{ version: '1.2.3' }],
       });
       expect(
         await resolveConstraint({ toolName: 'composer', constraint: 'whoops' })
       ).toBe('1.2.3');
     });
+
+    it.each`
+      version         | expected
+      ${'^3.9.0'}     | ${'3.10.4'}
+      ${'^3.9'}       | ${'3.10.4'}
+      ${'3.9.*'}      | ${'3.9.1'}
+      ${'3.9.1'}      | ${'3.9.1'}
+      ${'>3.8,<3.10'} | ${'3.9.1'}
+      ${'==3.9.1'}    | ${'3.10.4'}
```

Comment:
this case needs custom python versioning, which needs more time and isn't commonly used